internal/server/singleprocess/service_runner.go

```diff
@@ -110,22 +109,65 @@ func (s *service) RunnerConfig(
 		}
 	}()
 
+	// If this is an ODR runner, then we query the job it is waiting for
+	// in order to build up other information about this runner such as the
+	// project/app scope, workspace, etc.
+	//
+	// It is REQUIRED that an ODR has its target job queued BEFORE the
+	// ODR is launched. If we can't find a job, we error and exit which
+	// will also exit the runner.
+	var job *pb.Job
+	if record.Odr {
+		// Get a job assignment for this runner, non-blocking
+		sjob, err := s.state.JobAssignForRunner(ctx, record, false)
+		if err != nil {
+			return err
+		}
+		if sjob == nil {
+			return status.Errorf(codes.FailedPrecondition,
+				"no pending job for this on-demand runner. A pending job "+
+					"must be registered prior to registering the runner.")
+		}
+
+		// Nack the job cause we're not going to take it ourselves,
+		// the runner will later with JobStream.
+		if _, err := s.state.JobAck(sjob.Id, false); err != nil {
+			return err
+		}
```

Comment:
Should we have a `JobPeekForRunner` that doesn't perform the assignment?

Reply:
Didn't mean this to be a change request, just a comment.

Reply:
Yeah, this is probably a better idea.

Reply:
Done